src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1766 @@
+
+        # repeat same values for Batch_size and Num_Attn_Heads
+        offsets = offsets.repeat(batch_size, self.num_attention_heads, 1, 1)
+        offset_buckets = self._merge_seq_length_dims(buckets + offsets, self.num_attention_heads)
+
+        return offset_buckets
+
+    def _get_sorted_bucket_idx_and_undo_sorted_bucket_idx(self, sequence_length, buckets, num_hashes):
+        batch_size = buckets.shape[0]
+
+        orig_indices = torch.arange(num_hashes * sequence_length, device=buckets.device)
+        orig_indices = orig_indices.repeat(batch_size, self.num_attention_heads, 1)
+
+        # scale buckets
+        scaled_buckets = sequence_length * buckets + (orig_indices % sequence_length)
+
+        # remove gradient
+        scaled_buckets = scaled_buckets.detach()
+
+        # Hash-based sort
+        sorted_bucket_idx = torch.argsort(scaled_buckets, dim=-1)
+        undo_sorted_bucket_idx = torch.argsort(sorted_bucket_idx, dim=-1)
+
+        # remove gradient
+        sorted_bucket_idx = sorted_bucket_idx.detach()
+        undo_sorted_bucket_idx = undo_sorted_bucket_idx.detach()
+
+        return sorted_bucket_idx, undo_sorted_bucket_idx
+
+    def _set_num_buckets(self, sequence_length):
+        # recommended `num_buckets` from paper
+        num_buckets = 2 * sequence_length // self.chunk_length
+
+        # factorize `num_buckets` if `num_buckets` becomes too large
+        num_buckets_limit = max(int((self.max_position_embeddings // self.chunk_length) ** (0.5)), self.chunk_length,)
+        if num_buckets > 2 * num_buckets_limit:
+            num_buckets = [num_buckets_limit, num_buckets // num_buckets_limit + 1]
+
+        logger.warning("config.num_buckets is not set. Setting config.num_buckets to {}...".format(num_buckets))
+        self.num_buckets = num_buckets
+
+    def _attend(
+        self, query_vectors, key_vectors, value_vectors, sorted_bucket_idx, attention_mask, head_mask,
+    ):
+        key_vectors = self._look_adjacent(key_vectors, self.num_chunks_before, self.num_chunks_after)
+        value_vectors = self._look_adjacent(value_vectors, self.num_chunks_before, self.num_chunks_after)
+
+        # get logits and dots
+        query_key_dots = torch.matmul(query_vectors, key_vectors.transpose(-1, -2))
+
+        # free memory
+        del query_vectors, key_vectors
+
+        query_bucket_idx = self._split_seq_length_dim_to(
+            sorted_bucket_idx, -1, self.chunk_length, self.num_attention_heads
+        )
+        key_value_bucket_idx = self._look_adjacent(query_bucket_idx, self.num_chunks_before, self.num_chunks_after)
+
+        # get correct mask values depending on precision
+        if query_key_dots.dtype == torch.float16:
+            self_mask_value = self.self_mask_value_float16
+            mask_value = self.mask_value_float16
+        else:
+            self_mask_value = self.self_mask_value_float32
+            mask_value = self.mask_value_float32
```

Comment:
I need these special value for the integration tests and because of the rare cases when self_mask is the highest value (and thus has to be higher than mask)